openssl-dynamic/src/main/c/ssl.c

```diff
@@ -2086,6 +2086,117 @@ TCN_IMPLEMENT_CALL(void, SSL, setTlsExtHostName)(TCN_STDARGS, jlong ssl, jstring
     TCN_FREE_CSTRING(hostname);
 }
 
+TCN_IMPLEMENT_CALL(jobjectArray, SSL, authenticationMethods)(TCN_STDARGS, jlong ssl) {
+    SSL *ssl_ = J2P(ssl, SSL *);
+    const STACK_OF(SSL_CIPHER) *ciphers = NULL;
+    int len;
+    int i;
+    jobjectArray array;
+
+    if (ssl_ == NULL) {
+        tcn_ThrowException(e, "ssl is null");
+        return 0;
+    } else {
+        UNREFERENCED(o);
+
+        ciphers = SSL_get_ciphers(ssl_);
+        len = sk_num((_STACK*) ciphers);
+
+        array = (*e)->NewObjectArray(e, len, stringClass, NULL);
+
+        for (i = 0; i < len; i++) {
+            (*e)->SetObjectArrayElement(e, array, i,
+                (*e)->NewStringUTF(e, SSL_cipher_authentication_method((SSL_CIPHER*) sk_value((_STACK*) ciphers, i))));
+        }
+        return array;
+    }
+}
+
+TCN_IMPLEMENT_CALL(void, SSL, setCertificateBio)(TCN_STDARGS, jlong ssl,
+                                                         jlong cert, jlong key,
+                                                         jstring password)
+{
+    SSL *ssl_ = J2P(ssl, SSL *);
+    BIO *cert_bio = J2P(cert, BIO *);
+    BIO *key_bio = J2P(key, BIO *);
+    EVP_PKEY* pkey;
+    X509* xcert;
+    tcn_pass_cb_t* cb_data;
+    TCN_ALLOC_CSTRING(password);
+    char err[256];
+
+    UNREFERENCED(o);
+    TCN_ASSERT(ssl != NULL);
+
+    if (J2S(password)) {
+        cb_data = &tcn_password_callback;
+        strncpy(cb_data->password, J2S(password), SSL_MAX_PASSWORD_LEN);
+        cb_data->password[SSL_MAX_PASSWORD_LEN-1] = '\0';
+    }
+    if (key <= 0)
+        key = cert;
+    if (cert <= 0 || key <= 0) {
+        tcn_Throw(e, "No Certificate file specified or invalid file format");
+        goto cleanup;
+    }
+
+    if ((pkey = load_pem_key_bio(cb_data, key_bio)) == NULL) {
+        ERR_error_string_n(ERR_get_error(), err, 256);
```

Comment:
consider having a `const` or `#define` for the "commonly used error string length"